File: sesac7_flutter_recipe_app/lib/presentation/component/dialog/rating_dialog.dart
import 'package:flutter/material.dart';
import 'package:flutter_recipe_app/core/enum/rating_type.dart';
import 'package:flutter_recipe_app/presentation/component/button/rating_dialog_button.dart';
import 'package:flutter_recipe_app/ui/app_colors.dart';
import 'package:flutter_recipe_app/ui/text_styles.dart';

class RatingDialog extends StatefulWidget {
  final String dialogTitle;
  final String dialogButtonTitle;
  final RatingType currentRatingType;
  final Function(RatingType ratingType) onClickDissmissButton;

  const RatingDialog({
    super.key,
    required this.dialogTitle,
    required this.dialogButtonTitle,
    required this.currentRatingType,
    required this.onClickDissmissButton,
  });

  @override
  State<RatingDialog> createState() => _RatingDialogState();
}

class _RatingDialogState extends State<RatingDialog> {
  late RatingType _selectedRatingType;

  @override
  void initState() {
    super.initState();
    _selectedRatingType = widget.currentRatingType;
  }

  Icon _getRatingIcon(RatingType ratingType) =>
      ratingType.getIntValue() <= _selectedRatingType.getIntValue()
      ? Icon(Icons.star, color: AppColors.rating, size: 20.0)
      : Icon(Icons.star_border_outlined, color: AppColors.rating, size: 20.0);

  @override
  Widget build(BuildContext context) {
    return AlertDialog(
      insetPadding: const EdgeInsets.symmetric(horizontal: 15.0),
      backgroundColor: AppColors.white,
      shape: RoundedRectangleBorder(borderRadius: BorderRadius.circular(8.0)),
      elevation: 3.0,
      shadowColor: Colors.black,
      title: Text(
        widget.dialogTitle,
        style: TextStyles.ratingDialogTitle,
        textAlign: TextAlign.center,
      ),
      titlePadding: const EdgeInsets.only(top: 10.0),
      content: Row(
        mainAxisAlignment: MainAxisAlignment.center,
        mainAxisSize: MainAxisSize.min,
        spacing: 10.0,
        children: [
          for (final ratingType in [
            RatingType.gradeOne,
            RatingType.gradeTwo,
            RatingType.gradeThree,
            RatingType.gradeFour,
            RatingType.gradeFive,
          ])
            GestureDetector(
              onTap: () {
                setState(() {
                  _selectedRatingType = ratingType;
                });
              },
              child: _getRatingIcon(ratingType),
            ),
        ],
      ),
      contentPadding: const EdgeInsets.symmetric(
        vertical: 5.0,
      ),
      actions: <Widget>[
        Center(
          child: RatingDialogButton(
            buttonTitle: 'Send',
            buttonActive: _selectedRatingType.getIntValue() > 0,
            onTap: () {
              if (_selectedRatingType != RatingType.gradeZero) {
                widget.onClickDissmissButton(_selectedRatingType);
              }
            },
          ),
        ),
      ],
      actionsPadding: const EdgeInsets.only(bottom: 10.0),
    );
  }
}

File: compose_recipe/app/src/main/java/com/example/recipeapp/presentation/component/dialog/RatingDialog.kt
package com.example.recipeapp.presentation.component.dialog

import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Star
import androidx.compose.material.icons.outlined.StarOutline
import androidx.compose.material3.BasicAlertDialog
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.dp
import androidx.compose.ui.window.DialogProperties
import com.example.recipeapp.core.enum.RatingType
import com.example.recipeapp.presentation.component.button.RatingDialogButton
import com.example.recipeapp.ui.theme.Rating
import com.example.recipeapp.ui.theme.White
import com.example.recipeapp.ui.theme.ratingDialogTitle

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun RatingDialog(
    modifier: Modifier = Modifier,
    onSendButtonClick: () -> Unit
) {
    var currentRatingType by remember { mutableStateOf(RatingType.GRADE_ZERO) }

    BasicAlertDialog(
        modifier = modifier
            .width(170.dp)
            .height(91.dp)
            .background(
                color = White,
                shape = RoundedCornerShape(
                    size = 10.dp
                )
            ),
        onDismissRequest = {},
        properties = DialogProperties(usePlatformDefaultWidth = false)
    ) {
        Column {
            Text(
                text = "Rate recipe",
                modifier = Modifier
                    .fillMaxWidth()
                    .padding(top = 10.dp, bottom = 5.dp),
                style = ratingDialogTitle,
                textAlign = TextAlign.Center
            )
            Box(
                modifier = Modifier
                    .fillMaxWidth(),
                contentAlignment = Alignment.Center
            ) {
                Row(
                    modifier = Modifier
                        .padding(vertical = 2.dp),
                    horizontalArrangement = Arrangement.spacedBy(10.dp)
                ) {
                    for (i in listOf<RatingType>(
                        RatingType.GRADE_ONE,
                        RatingType.GRADE_TWO,
                        RatingType.GRADE_THREE,
                        RatingType.GRADE_FOUR,
                        RatingType.GRADE_FIVE
                    )) {
                        Icon(
                            if (i.value <= currentRatingType.value)
                                Icons.Default.Star
                            else
                                Icons.Outlined.StarOutline,
                            modifier = Modifier
                                .size(20.dp)
                                .clickable {
                                    currentRatingType = i
                                },
                            tint = Rating,
                            contentDescription = null
                        )
                    }
                }
            }
            Box(
                modifier = Modifier
                    .fillMaxWidth()
                    .padding(top = 5.dp, bottom = 10.dp),
                contentAlignment = Alignment.Center
            ) {
                RatingDialogButton(
                    buttonActive = currentRatingType != RatingType.GRADE_ZERO,
                    onTap = onSendButtonClick
                )
            }
        }
    }
}

@Preview
@Composable
fun RatingDialogPreview() {
    RatingDialog(
        onSendButtonClick = {

        }
    )
}